pkg/kubernetes/deployment/appsv1/v1alpha1/deployment.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2019 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	api_apps_v1 "k8s.io/api/apps/v1"
+)
+
+// predicate abstracts conditional logic w.r.t the deployment instance
+//
+// NOTE:
+// predicate is a functional approach versus traditional approach to mix
+// conditions such as *if-else* within blocks of business logic
+//
+// NOTE:
+// predicate approach enables clear separation of conditionals from
+// imperatives i.e. actions that form the business logic
+type predicate func(*deploy) bool
+
+// deployBuildOption is a typed function that abstracts anykind of operation
+// against the provided deployment instance
+//
+// This is the basic building block to create functional operations
+// against the deployment instance
+type deployBuildOption func(*deploy)
+
+// rolloutStatus  is a typed function that abstracts status message formation logic
+type rolloutStatus func(*deploy) string
+
+// deploy is a wrapper over aapi_apps_v1.Deployment
+type deploy struct {
+	object *api_apps_v1.Deployment // kubernetes deployment instance
+	checks []predicate             // predicate list for deploy
+}
+
+// predicateKey is wrapper over string. It is used to get predicate function and status msg.
+type predicateKey string
+
+// rolloutStatuses contains a group of status message for each predicate checks.
+// It useses predicateKey as key.
+var rolloutStatuses = map[predicateKey]rolloutStatus{
+	// ProgressDeadlineExceededPK refer to rolloutStatus for predicate IsProgressDeadlineExceeded.
+	ProgressDeadlineExceededPK: func(d *deploy) string {
+		return "Deployment exceeded its progress deadline"
+	},
+	// OlderReplicaActivePK refer to rolloutStatus for predicate IsOlderReplicaActive.
+	OlderReplicaActivePK: func(d *deploy) string {
+		if d.object.Spec.Replicas != nil {
+			return fmt.Sprintf("Waiting for deployment rollout to finish: %d out of %d new replicas have been updated",
+				d.object.Status.UpdatedReplicas, *d.object.Spec.Replicas)
+		}
+		return "Waiting for deployment rollout to finish: some older replicas have been updated"
```

Comment:
will it be better to do below:
```go
func(d *deploy) string {
  if d.object.Spec.Replicas == nil {
    return "Waiting for deployment rollout to finish: some older replicas have been updated"
  }
  return fmt.Sprintf("Waiting for deployment rollout to finish: %d out of %d new replicas have been updated",
				d.object.Status.UpdatedReplicas, *d.object.Spec.Replicas)
}
```

Reply:
I am not so sure about the messages.
Can we write below instead: 
`"old_replica_active: deployment rollout in-progress"`
&
`"old_replica_active: deployment rollout in-progress: %d of %d replicas are updated"`

I am not too sure if `old`, `new` terms should be used.
I felt it will make sense if we avoid lengthy sentences as:
1/ we may not be so sure of the description.
2/ it may confuse readers when so many terms/info is being pushed

NOTE: Above advice differs based on case to case basis.

Reply:
We also need to do a dry run by comparing all the messages & verify if they make sense when seen besides each other.

Reply:
If above messaging makes sense, please apply it across all

Reply:
done please check it once.